### ISSUE TITLE
CPT will not compile on 3.7 and 3.8 + wheel package will not import

### ISSUE THREAD

**Issue by catchthemonster** [bluesheeptoken/CPT#77]
Hi,
following directions both ways:
pip3 install cython && python3.7(8) setup.py install throws tons of erroirs:
Error compiling Cython file:
```
------------------------------------------------------------
...
from cpt.prediction_tree cimport PredictionTree
^
------------------------------------------------------------

cpt/cpt.pxd:1:0: 'cpt/prediction_tree.pxd' not found

Error compiling Cython file:
------------------------------------------------------------
...
from cpt.prediction_tree cimport PredictionTree
^
------------------------------------------------------------

cpt/cpt.pxd:1:0: 'cpt/prediction_tree/PredictionTree.pxd' not found

Error compiling Cython file:
------------------------------------------------------------
...
from cpt.prediction_tree cimport PredictionTree
from cpt.alphabet cimport Alphabet
....


installing cpt 1.2.1 module by pip3 succeeds but import will not work

rom cpt import CPT
Traceback (most recent call last):
  File "<input>", line 1, in <module>
ImportError: cannot import name 'CPT' from 'cpt' (unknown location)
from cpt import Cpt
Traceback (most recent call last):
  File "<input>", line 1, in <module>
ImportError: cannot import name 'Cpt' from 'cpt' (unknown location)
```
How do you get this module to properly install...?

**Comment by bluesheeptoken** [comment]
Hello,
thanks for reporting the issue.
I have published wheels for all platforms, including python3.8. They should work properly.

Can you try a `pip install cpt==1.2.1` and tell me if it is working ?

If not, can you telling me the platform on which you try to install ?

**Comment by catchthemonster** [comment]
sasha@venus:/cube/api/py-3.8.0/bin$ ./pip3 install cpt==1.2.1 
Requirement already satisfied: cpt==1.2.1 in /cube/api/py-3.8.0/lib/python3.8/site-packages (1.2.1)
sasha@venus:/cube/api/py-3.8.0/bin$ ./python3
python3           python3.8         python3.8-config  python3-config    
sasha@venus:/cube/api/py-3.8.0/bin$ ./python3
Python 3.8.0 (default, Nov 13 2019, 17:08:05) 
[GCC 9.2.1 20191109] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from cpt import Cpt
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'Cpt' from 'cpt' (unknown location)
>>> from cpt import CPT
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'CPT' from 'cpt' (unknown location)

This is on debian buster or 10.4
ls /cube/api/py-3.8.0/lib/python3.8/site-packages/cpt
alphabet.cpython-38-x86_64-linux-gnu.so  cpt.cpython-38-x86_64-linux-gnu.so

Remember this is custom installation of python 3, meaning non standard path. I also installed cPython...
To me it is not clear where the Python classes are, all what I see are dynamic libraries ...

/cube/api/py-3.8.0/lib/python3.8/site-packages# ls cpt*
cpt:
alphabet.cpython-38-x86_64-linux-gnu.so  cpt.cpython-38-x86_64-linux-gnu.so

cpt-1.2.1.dist-info:
INSTALLER  LICENSE  METADATA  RECORD  top_level.txt  WHEEL

cpt.libs:
libgomp-f7e03b3e.so.1.0.0

**Comment by bluesheeptoken** [comment]
This is really weird ...

So if I understood well, you tried first to install via sources. It obviously did not work but pip considers it installed, so a `pip install` does not work either.

I might be stubborned, can you try a `pip uninstall cpt && pip install cpt` ? I published wheels for `manylinux` which should be compatible for Debian like distributions.

---

The reason you are seeing only dynamic libraries is because I only expose `Alphabet` and `Cpt` to the end users. The other classes are written in CXX and should not be exposed.

Basically the pipeline is this one => Cython files to cxx, cxx to dynamic libraries with Cython generated cxx having a wrapper for python.

---

Can you try a  `pip3 uninstall cpt && pip3 install cpt` . Uninstall to make sure you do not have any files left from the source installation.

If that does not work, I will try to have a deeper look at it and try to reproduce it with your python installation (if you can link me how you installed it)

**Comment by catchthemonster** [comment]
hi, 
uninstalled it
then:
./pip3 install cpt
Collecting cpt
  Using cached cpt-1.2.1-cp38-cp38-manylinux2010_x86_64.whl (757 kB)
Installing collected packages: cpt
Successfully installed cpt-1.2.1

sasha@venus:/cube/api/py-3.8.0/bin$ ./python3
Python 3.8.0 (default, Nov 13 2019, 17:08:05) 
[GCC 9.2.1 20191109] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from cpt impot Cpt
  File "<stdin>", line 1
    from cpt impot Cpt
             ^
SyntaxError: invalid syntax
>>> from cpt import Cpt
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'Cpt' from 'cpt' (unknown location)
>>> from cpt import CPT
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'CPT' from 'cpt' (unknown location)

tghen new version 3.8.3
sudo apt update
sudo apt upgrade
sudo apt install wget build-essential
sudo apt install libssl-dev zlib1g-dev libncurses5-dev libncursesw5-dev libreadline-dev libsqlite3-dev
sudo apt install libgdbm-dev libdb5.3-dev libbz2-dev libexpat1-dev liblzma-dev libffi-dev uuid-dev tk-dev
pwd
/cube/build/Python-3.8.3
./configure --enable-optimizations --prefix=/cube/api/py-3.8.3
sudo make -j 16
sudo make install

Collecting setuptools
Collecting pip
Installing collected packages: setuptools, pip
Successfully installed pip-19.2.3 setuptools-41.2.0

then
cd /cube/api/py-3.8.3/bin
.sasha@venus:/cube/api/py-3.8.3/bin$ sudo ./pip3 install cpt
Collecting cpt
  Downloading https://files.pythonhosted.org/packages/4b/ea/9396a1e74456568c3fc76790f61358f6362b77a2132fa84d879dc6adc351/cpt-1.2.1-cp38-cp38-manylinux2010_x86_64.whl (757kB)
     |████████████████████████████████| 757kB 2.8MB/s 
Installing collected packages: cpt
Successfully installed cpt-1.2.1
WARNING: You are using pip version 19.2.3, however version 20.1.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
sasha@venus:/cube/api/py-3.8.3/bin$ sudo ./pip3 install --upgrade pip
Collecting pip
  Downloading https://files.pythonhosted.org/packages/43/84/23ed6a1796480a6f1a2d38f2802901d078266bda38388954d01d3f2e821d/pip-20.1.1-py2.py3-none-any.whl (1.5MB)
     |████████████████████████████████| 1.5MB 2.7MB/s 
Installing collected packages: pip
  Found existing installation: pip 19.2.3
    Uninstalling pip-19.2.3:
      Successfully uninstalled pip-19.2.3
Successfully installed pip-20.1.1

finally:
sasha@venus:/cube/api/py-3.8.3/bin$ ./python3
Python 3.8.3 (default, Jun 26 2020, 08:51:19) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from cpt import Cpt
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'Cpt' from 'cpt' (unknown location)
>>> from cpt import CPT
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'CPT' from 'cpt' (unknown location)


So in conclusion, simply does not work or it can't find Cpt or CPT entry point...

>> import cpt
>>> dir(cpt)
['__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__']

**Comment by bluesheeptoken** [comment]
In the last release I tried to set the endpoint at `cpt`. But this might not work well ...

Can you try a `from cpt.cpt import Cpt` ?

I feel like the installation is now ok, but my endpoint exposure is wrong. `from cpt.cpt import Cpt` should directly load the class.

**Comment by catchthemonster** [comment]
ha, that worked
and example did work too!
So we know now, thanks bluessheeptoken, which is really funny ID.
Thanks much
Please let us know if import changes in the future!!

**Comment by bluesheeptoken** [comment]
Thank you very much for pointing the issue ! I should definitely fix it this week end

> So we know now, thanks bluessheeptoken, which is really funny ID.
Thanks !

**Comment by catchthemonster** [comment]
i can help you with documentation if you wish!
if you give me an access to modify, I will fix documentation and as I go with options and different ways of implementing I can write up additional info. You will always approve it before a post ...
No worries, my wife is academic so she can correct my write ups if need be...
Ether way is fine, I am just offering ...

**Comment by bluesheeptoken** [comment]
Thanks, I appreciate !
I invited you to collaborate

**Comment by catchthemonster** [comment]
ok will help!
One more thing that would be very helpful for my project is to create a python wrapper that could load files into train and test for prediction.
let me know if youi want me to write in python wrapper that can present  these payloads to cpt

**Comment by bluesheeptoken** [comment]
I think CPT should not have the responsability to load files in memory.

However, you can probably use [train_test_split](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html) from sklearn package. 

CPT should also be compatible with sklearn tools, cf: https://cpt.readthedocs.io/en/latest/example.html#sklearn-example

However if you find any useful methods worth to be in a module, we can still see to create a separate module !
